File: 227. Basic Calculator II/calculate.js
/**
 * @param {string} s
 * @return {number}
 */
const calculate = (s) => {
  const stack = [];
  let num = 0;
  let prevOperator = "+";

  for (let i = 0; i < s.length; i++) {
    if (!isNaN(s[i]) && s[i] !== " ") {
      num = num * 10 + parseInt(s[i]);
    }

    if ((isNaN(s[i]) && s[i] !== " ") || i === s.length - 1) {
      switch (prevOperator) {
        case "+":
          stack.push(num);
          break;
        case "-":
          stack.push(-num);
          break;
        case "*":
          stack.push(stack.pop() * num);
          break;
        case "/":
          stack.push(Math.trunc(stack.pop() / num));
          break;
        default:
          break;
      }
      num = 0;
      prevOperator = s[i];
    }
  }

  return stack.reduce((acc, cur) => acc + cur, 0);
};


// s = "3+2*2";
s = "3/2";
console.log("calculate", calculate(s)); // 7

File: 227. Basic Calculator II/calculate.py
class Solution:
  def calculate(self, s: str) -> int:
    stack = []
    num = 0
    prevOperator = '+'

    for i in range(len(s)):
      if s[i].isdigit():
        num = num * 10 + int(s[i])

      if (not s[i].isdigit() and s[i] != ' ') or i == len(s) - 1:
        if prevOperator == '+':
          stack.append(num)
        elif prevOperator == '-':
          stack.append(-num)
        elif prevOperator == '*':
          stack.append(stack.pop() * num)
        elif prevOperator == '/':
          stack.append(int(stack.pop() / num))

        prevOperator = s[i]
        num = 0

    return sum(stack)
  

solution = Solution()
# s = "3+2*2"
s = " 3/2 "
print('calculate: ', solution.calculate(s)) # 7
